backend/lib/edgehog/containers/changes/destroy_related_releases.ex

```diff
@@ -0,0 +1,50 @@
+#
+# This file is part of Edgehog.
+#
+# Copyright 2025 SECO Mind Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+defmodule Edgehog.Containers.Changes.DestroyRelatedReleases do
+  @moduledoc false
+  use Ash.Resource.Change
+
+  @impl Ash.Resource.Change
+  def change(changeset, _opts, %{tenant: tenant}) do
+    application = Ash.load!(changeset.data, [:releases], tenant: tenant)
+
+    undeletable_releases =
+      Enum.reject(application.releases, fn release ->
+        %{valid?: valid} = Ash.Changeset.for_destroy(release, :destroy)
+        valid
+      end)
+
+    if undeletable_releases == [] do
+      Enum.each(application.releases, fn release ->
+        Ash.destroy!(release, tenant: tenant)
+      end)
+
+      changeset
```

Comment:
This has to be wrapped in an `after_action`, we do not want to delete all the releases and then something goes wrong with the deletion of the application or vice versa, leaving in an inconsistent state

```suggestion
Ash.Changeset.after_action(changeset, fn _changeset, _application ->
  with {:ok, _} <- result do 
      Enum.each(application.releases, fn release ->
        Ash.destroy!(release, tenant: tenant)
      end)
  end
  result
end)
```